cmd/encryption-v1.go

```diff
@@ -1157,3 +1160,140 @@ func DecryptObjectInfo(info *ObjectInfo, headers http.Header) (encrypted bool, e
+	var (
+		encryption encrypt.ServerSide
+		opts       ObjectOptions
+	)
+	if GlobalGatewaySSE != nil {
+		for _, k := range GlobalGatewaySSE {
+			if k == GatewaySSEC && crypto.SSEC.IsRequested(r.Header) {
+				derivedKey, err := deriveClientKey(r.Header, crypto.SSECKey, bucket, object)
+				if err != nil {
+					return opts, err
+				}
+				encryption, err = encrypt.NewSSEC(derivedKey[:])
+				if err != nil {
+					return opts, err
+				}
+				return ObjectOptions{ServerSideEncryption: encryption}, nil
+			}
+		}
+	}
+	// default case of passing encryption headers to backend
+	return extractEncryptionOption(r.Header, false, nil)
+}
+
+// get ObjectOptions for PUT calls from encryption headers
+func putEncryptionOpts(r *http.Request, bucket, object string, metadata map[string]string) (opts ObjectOptions, err error) {
+	if GlobalGatewaySSE != nil {
+		for _, k := range GlobalGatewaySSE {
+			// In the case of multipart custom format, the metadata needs to be checked in addition to header to see if it
+			// is SSE-S3 encrypted, primarily because S3 protocol does not require SSE-S3 headers in PutObjectPart calls
+			if k == GatewaySSES3 && (crypto.S3.IsRequested(r.Header) || crypto.S3.IsEncrypted(metadata)) {
+				opts = ObjectOptions{ServerSideEncryption: encrypt.ServerSide(encrypt.NewSSE())}
+			}
+			if k == GatewaySSEC && crypto.SSEC.IsRequested(r.Header) {
+				return getEncryptionOpts(r, bucket, object)
+			}
+		}
+	}
+	// default case of passing encryption headers to backend
+	return extractEncryptionOption(r.Header, false, metadata)
+}
+
+// get ObjectOptions for Copy calls for encryption headers provided on the target side
+func copyDstEncryptionOpts(r *http.Request, bucket, object string, metadata map[string]string) (opts ObjectOptions, err error) {
+	if GlobalGatewaySSE != nil {
+		for _, k := range GlobalGatewaySSE {
+			// In the case of multipart custom format, the metadata needs to be checked in addition to header to see if it
+			// is SSE-S3 encrypted, primarily because S3 protocol does not require SSE-S3 headers in CopyObjectPart calls
+			if k == GatewaySSES3 && (crypto.S3.IsRequested(r.Header) || crypto.S3.IsEncrypted(metadata)) {
+				opts = ObjectOptions{ServerSideEncryption: encrypt.ServerSide(encrypt.NewSSE())}
+			}
+			if k == GatewaySSEC && crypto.SSEC.IsRequested(r.Header) {
+				return getEncryptionOpts(r, bucket, object)
+			}
+		}
+	}
+	// default case of passing encryption headers to backend
+	return extractEncryptionOption(r.Header, false, nil)
+}
+
+// get ObjectOptions for Copy calls for encryption headers provided on the source side
+func copySrcEncryptionOpts(r *http.Request, bucket, object string) (ObjectOptions, error) {
+	var (
+		ssec encrypt.ServerSide
+		opts ObjectOptions
+	)
+	if GlobalGatewaySSE != nil {
+		for _, k := range GlobalGatewaySSE {
+			if k == GatewaySSEC && crypto.SSECopy.IsRequested(r.Header) {
+				derivedKey, err := deriveClientKey(r.Header, crypto.SSECopyKey, bucket, object)
```

Comment:
Same - use `crypto.SSECopy.ParseHTTP(...)` and pass key directly.

Reply:
done